Dotty deviation: failing to compile case class & package object with same name

```
case class FooBar(){}
package object FooBar{}
```
```
exception occured while indexing packageCase.scala
Exception in thread "main" scala.MatchError: PackageDef(Ident(FooBar),List(ModuleDef(package,Template(DefDef(<init>,List(),List(),TypeTree(EmptyTree),EmptyTree),List(),ValDef(_,EmptyTree,EmptyTree),List(EmptyTree))))) (of class dotty.tools.dotc.ast.Trees$PackageDef)
	at dotty.tools.dotc.typer.Namer$$anonfun$mergeCompanionDefs$1$4.apply(Namer.scala:408)
	at dotty.tools.dotc.typer.Namer$$anonfun$mergeCompanionDefs$1$4.apply(Namer.scala:405)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:778)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:777)
	at dotty.tools.dotc.typer.Namer.mergeCompanionDefs$1(Namer.scala:405)
	at dotty.tools.dotc.typer.Namer.index(Namer.scala:422)
	at dotty.tools.dotc.typer.Namer.indexExpanded(Namer.scala:380)
	at dotty.tools.dotc.typer.Namer.index(Namer.scala:371)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$enterSyms$1.apply$mcV$sp(FrontEnd.scala:35)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:18)
	at dotty.tools.dotc.typer.FrontEnd.enterSyms(FrontEnd.scala:33)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$2.apply(FrontEnd.scala:51)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$2.apply(FrontEnd.scala:51)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:51)
	at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:50)
	at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:47)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply$mcV$sp(Run.scala:47)
	at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:42)
	at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:42)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
	at dotty.tools.dotc.Run.compileSources(Run.scala:41)
	at dotty.tools.dotc.Run.compile(Run.scala:32)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
	at dotty.tools.dotc.Main$.doCompile(Main.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:31)
	at dotty.tools.dotc.Driver.main(Driver.scala:40)
	at dotty.tools.dotc.Main.main(Main.scala)
```

failure in typer:
https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/typer/Typer.scala#L966
```
packageCase.scala:2: error: object FooBar is not a package
package object FooBar{}
                     ^
one error found
```

--- a/src/dotty/tools/dotc/typer/Implicits.scala
+++ b/src/dotty/tools/dotc/typer/Implicits.scala
@@ -402,7 +402,8 @@ trait Implicits { self: Typer =>
         || (to isRef defn.ObjectClass)
         || (to isRef defn.UnitClass)
         || (from.tpe isRef defn.NothingClass)
-        || (from.tpe isRef defn.NullClass)) NoImplicitMatches
+        || (from.tpe isRef defn.NullClass)
+        || (from.tpe eq NoPrefix)) NoImplicitMatches
     else
       try inferImplicit(to.stripTypeVar.widenExpr, from, from.pos)
       catch {

--- a/src/dotty/tools/dotc/typer/TypeAssigner.scala
+++ b/src/dotty/tools/dotc/typer/TypeAssigner.scala
@@ -8,6 +8,7 @@ import Scopes._, Contexts._, Constants._, Types._, Symbols._, Names._, Flags._, 
 import ErrorReporting._, Annotations._, Denotations._, SymDenotations._, StdNames._, TypeErasure._
 import util.Positions._
 import config.Printers._
+import NameOps._
 
 trait TypeAssigner {
   import tpd._
@@ -16,7 +17,11 @@ trait TypeAssigner {
    *  @param packageOk   The qualifier may refer to a package.
    */
   def qualifyingClass(tree: untpd.Tree, qual: Name, packageOK: Boolean)(implicit ctx: Context): Symbol = {
-    def qualifies(sym: Symbol) = sym.isClass && (qual.isEmpty || sym.name == qual)
+    def qualifies(sym: Symbol) =
+      sym.isClass && (
+          qual.isEmpty ||
+          sym.name == qual ||
+          sym.is(Module) && sym.name.stripModuleClassSuffix == qual)
     ctx.outersIterator.map(_.owner).find(qualifies) match {
       case Some(c) if packageOK || !(c is Package) =>
         c

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -1068,7 +1068,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     val packageContext =
       if (pkg is Package) ctx.fresh.setOwner(pkg.moduleClass).setTree(tree)
       else {
-        ctx.error(d"$pkg is not a package", tree.pos)
+        ctx.error(d"$pkg is already defined, cannot be a package", tree.pos)
         ctx
       }
     val stats1 = typedStats(tree.stats, pkg.moduleClass)(packageContext)

--- a/test/dotc/tests.scala
+++ b/test/dotc/tests.scala
@@ -155,6 +155,7 @@ class tests extends CompilerTest {
   @Test def neg_i0091_infpaths = compileFile(negDir, "i0091-infpaths", xerrors = 3)
   @Test def neg_i0248_inherit_refined = compileFile(negDir, "i0248-inherit-refined", xerrors = 4)
   @Test def neg_i0281 = compileFile(negDir, "i0281-null-primitive-conforms", xerrors = 3)
+  @Test def neg_i324 = compileFile(negDir, "i324", xerrors = 2)
   @Test def neg_i583 = compileFile(negDir, "i0583-skolemize", xerrors = 2)
   @Test def neg_i941 = compileFile(negDir, "i941", xerrors = 3)
   @Test def neg_finalSealed = compileFile(negDir, "final-sealed", xerrors = 2)

--- a/tests/neg/i324.scala
+++ b/tests/neg/i324.scala
@@ -1,0 +1,5 @@
+class O
+object O {
+   val x: this.type = OO.this
+   val y: O = OO.this
+}

--- a/tests/pos/i324.scala
+++ b/tests/pos/i324.scala
@@ -1,0 +1,7 @@
+class O
+object O {
+   val x: this.type = O.this
+}
+object O2 {
+   val x: this.type = O2.this
+}

